### PR TITLE
[OpenVINO backend] supporting inference for Gemma,  Mistral and GPT2 with ov backend

### PR DESCRIPTION
**Pull request by Mohamed-Ashraf273**
## Description of the change

As a part of my GSoC25 project to support inference with the openvino backend for `Gemma` , `Mistral` and `GPT-2`, 
This is my PR for supporting `Gemma` , `Mistral`  and `GPT-2` pipelines.
```python
import os
os.environ["KERAS_BACKEND"] = "openvino"
import keras_hub

model = keras_hub.models.GPT2CausalLM.from_preset(
    "gpt2_large_en", dtype="float16"
)
model.summary()
output = model.generate("Keras is ", max_length=20)
print("Generated text:", output)
```

## Reference

https://docs.openvino.ai/2025/index.html
https://keras.io/api/
https://keras.io/keras_hub/

## Checklist

- [x] I have added all the necessary unit tests for my change.
- [x] I have verified that my change does not break existing code and works with all backends (TensorFlow, JAX, and PyTorch).
- [x] My PR is based on the latest changes of the main branch (if unsure, rebase the code).
- [x] I have followed the Keras Hub [Model contribution guidelines](https://github.com/keras-team/keras-hub/blob/master/CONTRIBUTING_MODELS.md) in making these changes.
- [x] I have followed the Keras Hub [API design guidelines](https://github.com/keras-team/keras-hub/blob/master/API_DESIGN_GUIDE.md) in making these changes.
- [x] I have signed the [Contributor License Agreement](https://cla.developers.google.com/about).

### CODE REVIEW

File: keras_hub/src/models/gemma/gemma_causal_lm_test.py [hunk 1]
```diff
@@ -211,6 +237,10 @@ def test_all_presets(self):
                 input_data=self.input_data,
             )
 
+    @pytest.mark.skipif(
+        keras.config.backend() == "openvino",
+        reason="OpenVINO is for inference only",
```
Comment (mattdangerw) [inline keras_hub/src/models/gemma/gemma_causal_lm_test.py]: score is inference only, is there a reason this needs to be disabled?

File: keras_hub/src/utils/tensor_utils.py [hunk 2]
```diff
@@ -158,7 +158,8 @@ def convert_preprocessing_inputs(x):
         # If we have a string input, use tf.tensor.
         if isinstance(x, np.ndarray) and x.dtype.type is np.str_:
             return tf.convert_to_tensor(x)
-        x = ops.convert_to_tensor(x)
+        if keras.config.backend() != "openvino":
+            x = ops.convert_to_tensor(x)
```
Comment (mattdangerw) [inline keras_hub/src/utils/tensor_utils.py]: what does convert_to_tensor do on the openvino backend, should it just convert to numpy?

File: keras_hub/src/models/gemma/gemma_causal_lm_test.py [hunk 3]
```diff
@@ -89,6 +97,9 @@ def test_generate(self):
         causal_lm.preprocessor = None
         outputs = causal_lm.generate(prompt_ids, stop_token_ids=None)
         # Assert prompt is in output in token id space.
+        if keras.config.backend() == "openvino":
```
Comment (mattdangerw) [inline keras_hub/src/models/gemma/gemma_causal_lm_test.py]: Would extending this for our asserts mean we could ditch some of these switch cases? https://github.com/keras-team/keras-hub/blob/25c9062c5f9a25fade16094cd21d2545125168ec/keras_hub/src/tests/test_case.py#L21-L37

File: keras_hub/src/models/gemma/gemma_causal_lm_test.py [hunk 4]
```diff
@@ -149,6 +163,12 @@ def wrapper(*args, **kwargs):
             """Modify output logits to always favor end_token_id"""
             logits, hidden_states, cache = call_with_cache(*args, **kwargs)
             index = self.preprocessor.tokenizer.end_token_id
+            if keras.config.backend() == "openvino":
+                """Set all logits to a large negative number 
+                    to avoid NaNs produced by ov.einsum"""
+                logits = ops.ones_like(logits) * ops.convert_to_tensor(
```
Comment (mattdangerw) [inline keras_hub/src/models/gemma/gemma_causal_lm_test.py]: this seems suspicious, should we be fixing the code not the test? maybe we need a openvino fix before calling einsum in some places? what is failing?

File: keras_hub/src/models/gemma/gemma_causal_lm_test.py [hunk 5]
```diff
@@ -60,6 +64,10 @@ def test_causal_lm_basics(self):
             expected_output_shape=(2, 8, 11),
         )
 
+    @pytest.mark.skipif(
+        keras.config.backend() == "openvino",
+        reason="OpenVINO is for inference only",
```
Comment (mattdangerw) [inline keras_hub/src/models/gemma/gemma_causal_lm_test.py]: this is also inference only, no gradient descent. should this be enabled?

File: keras_hub/src/models/gemma/gemma_causal_lm.py [hunk 6]
```diff
@@ -196,22 +196,102 @@ def call_with_cache(
...
-        # Each decoder layer has a cache; we update them separately.
-        caches = []
-        for i, transformer_layer in enumerate(self.backbone.transformer_layers):
-            current_cache = cache[:, i, ...]
-            x, next_cache = transformer_layer(
-                x,
-                cache=current_cache,
-                cache_update_index=cache_update_index,
+
+        use_openvino = keras.config.backend() == "openvino"
```
Comment (mattdangerw) [inline keras_hub/src/models/gemma/gemma_causal_lm.py]: this seems quite hard to maintain and like something we should probably avoid. why do we need a pass that is so different than other backends? shouldn't the point of keras' multi-backend setup allow to just call layers normally?

File: keras_hub/src/models/causal_lm.py [hunk 7]
```diff
@@ -132,6 +137,113 @@ def make_generate_function(self):
             return self.generate_function
 
         self.generate_function = self.generate_step
+        if keras.config.backend() == "openvino":
+            import numpy as np
+            import openvino as ov
+            import openvino.runtime.opset14 as ov_opset
+            from keras.src.backend.openvino.core import OPENVINO_DTYPES
```
Comment (mattdangerw) [inline keras_hub/src/models/causal_lm.py]: we'd like to avoid private imports from keras if we can, they are hard to maintain if keras changes it's code structure.

File: keras_hub/src/models/causal_lm.py [hunk 8]
```diff
@@ -132,6 +137,113 @@ def make_generate_function(self):
             return self.generate_function
 
         self.generate_function = self.generate_step
+        if keras.config.backend() == "openvino":
```
Comment (mattdangerw) [inline keras_hub/src/models/causal_lm.py:135]: let's try to think if there's a way we can keep this resembling other backends a little more, this is fairly heavyweight.

**Review by mattdangerw — COMMENTED** [review]
Left some initial comments! But probably first question is around the changes to causal_lm and gemma_causal_lm. Why is this so backend specific? This is much more involved than changes for jax/torch/tensorflow

**Comment by Mohamed-Ashraf273** [comment]
> Left some initial comments! But probably first question is around the changes to causal_lm and gemma_causal_lm. Why is this so backend specific? This is much more involved than changes for jax/torch/tensorflow

hi @mattdangerw ,
I’ve been working on enabling inference for Gemma with OpenVINO by implementing the missing operations. The main challenge I ran into is that building the entire graph as a single model makes execution difficult (takes too long + RAM may overflow) , I'm still investigating why.
To address this, I introduced a subgraph-based approach by splitting the full graph at key points. I also added logic to store compiled subgraphs in CausalLM so they can be reused across generations instead of rebuilding them each time.

File: keras_hub/src/models/gemma/gemma_causal_lm_test.py (hunk 1 above)
Comment (Mohamed-Ashraf273) [reply keras_hub/src/models/gemma/gemma_causal_lm_test.py]: It needs `roll` operation, which is not supported, but I thought if I can make `generate` work for now, I'll enable this test.

File: keras_hub/src/utils/tensor_utils.py (hunk 2 above)
Comment (Mohamed-Ashraf273) [reply keras_hub/src/utils/tensor_utils.py]: At the end of the generate wrapper for OpenVINO, the output is already  NumPy. To avoid unnecessary reconversion in post_process, I disabled that step for the OpenVINO backend.

File: keras_hub/src/models/gemma/gemma_causal_lm_test.py (hunk 4 above)
Comment (Mohamed-Ashraf273) [reply keras_hub/src/models/gemma/gemma_causal_lm_test.py]: You are right, when I cancelled the splitting approach, it worked!

File: keras_hub/src/models/causal_lm.py (hunk 7 above)
Comment (Mohamed-Ashraf273) [reply keras_hub/src/models/causal_lm.py]: But I need them to make an OpenVINO model graph.

File: keras_hub/src/models/causal_lm.py (hunk 8 above)
Comment (Mohamed-Ashraf273) [reply keras_hub/src/models/causal_lm.py:135]: I’ve done my best to minimize backend-specific bias in the code.
I'd appreciate it if you could take another look.

**Comment by mattdangerw** [comment]
@Mohamed-Ashraf273 is there a way that we can land this without the subgraph approach?

We have a similar need in Jax at train time. Compilation times are much improved if you run a common transformer block in a compiled loop. So probably there is something to do here, but we'd really like to avoid our forward pass being a switch case on backend here. That will lead to maintenance hell.

So maybe let's try to land with the same approach as other backends for now, and see if there's a layer stacking/compilation reuse solution we can land as follow up?

**Comment by Mohamed-Ashraf273** [comment]
hi @mattdangerw 
I removed the subgraph approach + removed reusing part.
Now the model can be inferred with OpenVINO and pass all tests, we just need to think about how to optimize inference with large parameters and real weights without RAM overflow.
I'd appreiciate tit if you can take another look.
Thanks!

**Comment by Mohamed-Ashraf273** [comment]
@mattdangerw 
My PR is ready for review!

**Comment by Mohamed-Ashraf273** [comment]
@fchollet 
Can you take a look?

**Comment by Mohamed-Ashraf273** [comment]
@mattdangerw

File: keras_hub/src/models/causal_lm.py [hunk 9]
```diff
@@ -132,6 +132,123 @@ def make_generate_function(self):
...
+                "int8": ov.Type.i8,
+                "int16": ov.Type.i16,
+                "int32": ov.Type.i32,
+                "int64": ov.Type.i64,
+                "bfloat16": ov.Type.bf16,
+                "bool": ov.Type.boolean,
+                "float8_e4m3fn": ov.Type.f8e4m3,
+                "float8_e5m2": ov.Type.f8e5m2,
+                "string": ov.Type.string,
+            }
```
Comment (rkazants) [inline keras_hub/src/models/causal_lm.py]: can we import this dict from keras.src.backend.openvino.core?

**Comment by Mohamed-Ashraf273** [comment]
Hi @fchollet ,
I'd appreciate any feedback on my PR.
thanks

File: .github/workflows/actions.yml [hunk 10]
```diff
@@ -48,6 +51,10 @@ jobs:
       run: |
           pip install -r requirements.txt --progress-bar off
           pip install --no-deps -e "." --progress-bar off
+          if [[ "${{ matrix.backend }}" == "openvino" ]]; then
+            pip uninstall -y keras
+            pip install git+https://github.com/Mohamed-Ashraf273/keras.git@gsoc2025 --upgrade --force-reinstall --progress-bar off
```
Comment (divyashreepathihalli) [inline .github/workflows/actions.yml:56]: this should be removed for final merge

File: keras_hub/src/models/causal_lm.py (hunk 7 above)
Comment (divyashreepathihalli) [reply keras_hub/src/models/causal_lm.py]: Keras does not guarantee the stability of its src layout. A future refactoring in Keras could easily break this code. The goal should be to rely only on public Keras APIs.

Comment (divyashreepathihalli) [reply keras_hub/src/models/causal_lm.py]: maybe move the nested helper functions out of make_generate_function. They can be defined at the module level (e.g., in a keras_hub.src.backend.openvino_utils file if necessary) to make the core make_generate_function logic cleaner and more readable.

**Comment by Mohamed-Ashraf273** [comment]
@fchollet 
@mattdangerw 
@rkazants 
@divyashreepathihalli

**Comment by Mohamed-Ashraf273** [comment]
@mattdangerw 
@divyashreepathihalli
